bgpd/bgp_zebra.c

```diff
@@ -1393,7 +1393,17 @@ void bgp_zebra_announce(struct bgp_dest *dest, const struct prefix *p,
 
 		SET_FLAG(api.message, ZAPI_MESSAGE_OPAQUE);
 		api.opaque.length = strlen(aspath->str) + 1;
+		/*
+		 * Let's ensure that the as path string does not
+		 * go beyond the end of the buffer.  In this case
+		 * just hard limit the string to ZAPI_MESSAGE_OPAQUE_LENGTH
+		 * and put a `\0` terminating character, since
+		 * it's a string.
+		 */
+		if (api.opaque.length > ZAPI_MESSAGE_OPAQUE_LENGTH)
+			api.opaque.length = ZAPI_MESSAGE_OPAQUE_LENGTH;
 		memcpy(api.opaque.data, aspath->str, api.opaque.length);
+		api.opaque.data[ZAPI_MESSAGE_OPAQUE_LENGTH - 1] = 0;
```

Comment:
the decoder should do this too, right - ensure that the "string" is null-terminated?

Reply:
the decoder has no idea what is this data.  it's `opaque`.
if you send data down it will be properly handled in the appropriate call function in zebra_vty.c

Reply:
that isn't what it looks like to me - I just see the data memcpy'd, and then printed with '%s'

Reply:
The *decoder* in lib/zclient.c has absolutely no idea what the data is.  other than it must fit within the opaque data arrary.
I could pass down *anything* I want.  The code in zebra_vty.c has to handle it in a way that makes sense.  As I stated the decoder in lib/zclient.c has no idea about what the data is and it has no business verifying it like you asked it to.

Reply:
so ... maybe zebra needs to be able to protect itself based on the sender when decoding, in a similar way, since it's got vty code that's dependent on the sender?

Reply:
Read over both of your comments, read over the code in question in `zapi_msg.c:zread_route_add` and `zebra_vty.c:zebra_show_ip_route_opaque`. @mjstapp is saying that this code:
```
 1984         if (CHECK_FLAG(api.message, ZAPI_MESSAGE_OPAQUE)) {
  1985                 re->opaque = XMALLOC(MTYPE_OPAQUE,
  1986                                      sizeof(struct opaque) + api.opaque.length);
  1987                 re->opaque->length = api.opaque.length;      
  1988                 memcpy(re->opaque->data, api.opaque.data, re->opaque->length);  
  1989         }
```
should null terminate `re->opaque->data`. Presently the `memcpy` copies the null terminator when the sender is bgpd, who provides one. Mark is saying we should ensure this in all cases protects us from the case where the sender does not null terminate the string it encodes. @donaldsharp disagrees because in theory something other than a null terminated string could be passed in the opaque data, something which should not be null terminated, thus we cannot write logic that handles both cases without knowing what the opaque data contains. To which Mark replies that in that case perhaps Zebra should have code to null terminate when we know the data contains string, which we know based on the sender.

IMO doing that latter suggestion makes the data no longer opaque, so I disagree with that. Here's the change I think should be made:

```diff
diff --git a/zebra/zebra_vty.c b/zebra/zebra_vty.c
index f18d8fbb6..f74cfe41b 100644
--- a/zebra/zebra_vty.c
+++ b/zebra/zebra_vty.c
@@ -422,25 +422,24 @@ static void show_nexthop_detail_helper(struct vty *vty,
 static void zebra_show_ip_route_opaque(struct vty *vty, struct route_entry *re,
                                       struct json_object *json)
 {
+       char ostrbuf[re->opaque->length];
+       strlcpy(ostrbuf, re->opaque->data, sizeof(ostrbuf));
+
        if (!re->opaque)
                return;
 
        switch (re->type) {
        case ZEBRA_ROUTE_SHARP:
                if (json)
-                       json_object_string_add(json, "opaque",
-                                              (char *)re->opaque->data);
+                       json_object_string_add(json, "opaque", ostrbuf);
                else
-                       vty_out(vty, "    Opaque Data: %s",
-                               (char *)re->opaque->data);
+                       vty_out(vty, "    Opaque Data: %s", ostrbuf);
                break;
        case ZEBRA_ROUTE_BGP:
                if (json)
-                       json_object_string_add(json, "asPath",
-                                              (char *)re->opaque->data);
+                       json_object_string_add(json, "asPath", ostrbuf);
                else
-                       vty_out(vty, "    AS-Path: %s",
-                               (char *)re->opaque->data);
+                       vty_out(vty, "    AS-Path: %s", ostrbuf);
        default:
                break;
        }

```

Aside from that, imo for this PR the length check, length update, memcpy and null term should be replaced with
```c
api.opaque.length = MIN(ZAPI_MESSAGE_OPAQUE_LENGTH, strlen(aspath->str) + 1);
strlcpy(api.opaque.data, aspath->str, api.opaque.length);
```

Reply:
thanks for thinking this through, and I'm ok with whatever really. my eye saw the handling being done by the receiver, by zebra, and made me want zebra to do a bit more. maybe I'd like zebra to ignore these "OPAQUE" blobs from sources that don't have a formatter clause?

Reply:
updated code.

Reply:
I see no good way to ensure that zebra knows who is going to send it opaque data.  If an upper level protocol sends opaque data and there is no formatter, currently the data is just `held` until the route is released as that it can do nothing with the data.

Reply:
yes, I see that the data is held, and I don't think that's worth doing, but ... well, there it is. I don't think this is opaque data - zebra knows that the bgp string is an as-path, and I know that because it prints "AS-Path" - but I agree we should fix it and move on.

Reply:
> zebra knows that the bgp string is an as-path, and I know that because it prints "AS-Path"

So you ensure null termination when you print it, i.e. at the point where it makes sense to write code based on the known-in-advance buffer contents, by copying to a null terminated string buffer - the diff I provided, which is now implemented